scope.go

```diff
@@ -0,0 +1,622 @@
+package rcmgr
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/libp2p/go-libp2p-core/network"
+)
+
+// resources tracks the current state of resource consumption
+type resources struct {
+	limit Limit
+
+	nconnsIn, nconnsOut     int
+	nstreamsIn, nstreamsOut int
+	nfd                     int
+
+	memory int64
+}
+
+// A resourceScope can be a DAG, where a downstream node is not allowed to outlive an upstream node
+// (ie cannot call Done in the upstream node before the downstream node) and account for resources
+// using a linearized parent set.
+// A resourceScope can be a txn scope, where it has a specific owner; txn scopes create a tree rooted
+// at the owner (which can be a DAG scope) and can outlive their parents -- this is important because
+// txn scopes are the main *user* interface for memory management, and the user may call
+// Done in a txn scope after the system has closed the root of the txn tree in some background
+// goroutine.
+// If we didn't make this distinction we would have a double release problem in that case.
+type resourceScope struct {
+	sync.Mutex
+	done   bool
+	refCnt int
+
+	rc          resources
+	owner       *resourceScope   // set in transaction scopes, which define trees
+	constraints []*resourceScope // set in DAG scopes, it's the linearized parent set
```

Comment:
Can we call this `path` or `lineage`? I was fumbling to understand what a "constraint" was in this context because we had not introduced that concept.

Reply:
ok, i'll think of a better name.

Reply:
how about `edges`? They are edges in the DAG.

Reply:
renamed to edges.